File: C/Basic Programming/TicTacToe.c
#include <stdio.h>

void printBox(char *characters);
int conditionCheck(char *characters);

int main()
{
    int turn;
    char array[9] = {'1', '2', '3', '4', '5', '6', '7', '8', '9'};
    printf("\n   ***   Tic Tac Toe   ***   ");
    printf("\nPlayer1's Turn...");
    printBox(array);

    printf("\nType the number of position that you have to place your 'X' :- ");
    scanf("%d", &turn);

    array[turn - 1] = 'X';
    printBox(array);
    printf("\nPlayer1 placed his X at %d No. place.", turn);

    for (int i = 0; i < 4; i++)
    {
        printf("\nPlayer2's Turn...");
        printBox(array);
        printf("\nType the number of position that you have to place your 'O' :- ");
        scanf("%d", &turn);

        array[turn - 1] = 'O';
        printBox(array);
        printf("\nPlayer2 placed his O at %d No. place.", turn);

        if (conditionCheck(array))
        {
            printf("\n");
            printf("\nPlayer2 wins the match");
            printf("\n   *** Congratulations ***   ");
            printf("\n       - for Player2's Victory.");
            break;
        }
        else
        {
            printf("\nPlayer1's Turn...");
            printBox(array);
            printf("\nType the number of position that you have to place your 'X' :- ");
            scanf("%d", &turn);

            array[turn - 1] = 'X';
            printBox(array);
            printf("\nPlayer1 placed his X at %d No. place.", turn);

            if (conditionCheck(array))
            {
                printf("\n");
                printf("\nPlayer1 wins the match.");
                printf("\n   *** Congratulations ***   ");
                printf("\n      - for player1's Victory.");
                break;
            }
        }
    }
    if (conditionCheck(array) == 0)
    {
        printf("\n");
        printf("\nIt's a Tie match.");
        printf("\nwell played...");
        printf("\nBetter luck next time!");
    }
    return 0;
}
void printBox(char *characters)
{
    printf("\n");
    printf("\n");
    printf(" %c | %c | %c", characters[0], characters[1], characters[2]);
    printf("\n");
    printf("---+---+---");
    printf("\n");
    printf(" %c | %c | %c", characters[3], characters[4], characters[5]);
    printf("\n");
    printf("---+---+---");
    printf("\n");
    printf(" %c | %c | %c", characters[6], characters[7], characters[8]);
    printf("\n");
}
int conditionCheck(char *list)
{
    int win;
    if (list[0] == list[1] && list[1] == list[2])
    {
        win = 1;
    }
    else if (list[3] == list[4] && list[4] == list[5])
    {
        win = 1;
    }
    else if (list[6] == list[7] && list[7] == list[8])
    {
        win = 1;
    }
    else if (list[0] == list[3] && list[3] == list[6])
    {
        win = 1;
    }
    else if (list[1] == list[4] && list[4] == list[7])
    {
        win = 1;
    }
    else if (list[2] == list[5] && list[5] == list[8])
    {
        win = 1;
    }
    else if (list[0] == list[4] && list[4] == list[8])
    {
        win = 1;
    }
    else if (list[2] == list[4] && list[4] == list[6])
    {
        win = 1;
    }
    else
    {
        win = 0;
    }
    return win;
}

File: Python/Basic Python Programming/TicTacToe.py
board = [1, 2, 3, 4, 5, 6, 7, 8, 9]
tokens = ['X', 'O']

def printBoard():
    print(f'''
      {board [0]} | {board [1]} | {board [2]}
     ———+———+———
      {board [3]} | {board [4]} | {board [5]}
     ———+———+———
      {board [6]} | {board [7]} | {board [8]}
    
  
  ''')


def conditionCheck():
    conditions = [
        [0, 1, 2], [3, 4, 5], [6, 7, 8],    # Horizontal Checks
        [0, 3, 6], [1, 4, 7], [2, 5, 8],    # Vertical Checks
        [0, 4, 8], [2, 4, 6]                # DIagonal Checks
    ]
    win = False
    for i, j, k in conditions:
        if ((board[i]) == (board[j]) and (board[j]) == (board[k])):
            win = True
    return win

def isFull():
    for token in board:
        if token not in tokens:
            return False
    return True

def makeTurn(pos, token):
    if 0 <= pos <= 8:
        if board[pos] not in tokens:
            board[pos] = token
            return True
    return False

def game():
    player = 2
    print("***   Tic Tac Toe   ***   ")

    while not isFull():
        player = 1 if player == 2 else 2
        token = tokens[player-1]
        print(f"Player{player}'s turn...")
        printBoard()
        turn = int(input(f'''Type the number of position\nwhere you have to place your '{token}' : '''))

        if not makeTurn(turn-1, token):
            print("\nInvalid Move, Try again!!!\n\n")
            player = 1 if player == 2 else 2
            continue

        printBoard()
        print(f"Player{player} placed his {token} at {turn} No. place.")

        if conditionCheck():
            print(f'''Player{player} wins  match\n*** congratulations ***\n\t-for player{player}'s victory.''')
            break

    if not conditionCheck():
        print('''\n\nIt's a tie match\nWell played...\nBetter luck next time!''')

if __name__ == "__main__":
    game()
